BDDC demo (#5151)
firedrakeproject/firedrake: demos/bddc/bddc_poisson.py.rst

Cellwise Balancing Domain Decomposition by Constraints
======================================================

Contributed by `Pablo Brubeck <https://www.maths.ox.ac.uk/people/pablo.brubeckmartinez/>`_ and `Stefano Zampini <stefano.zampini@gmail.com>`_.

In this demo, we demonstrate how to configure and use Firedrake's Balancing Domain
Decomposition by Constraints (BDDC) preconditioners. BDDC is an 
advanced domain decomposition method suitable for massive parallelization of 
elliptic PDEs with great success for problems with
high anisotropy in the mesh or the coefficients.

Multielement subdomains in BDDC
-------------------------------

In standard domain decomposition methods, a global mesh is 
partitioned into subdomains consisting of several elements (typically via graph partitioners like METIS). 
Continuity across the subdomain edges/faces is removed and variationally enforced only at selected degrees of freedom (DoFs),

> refered to as `primal`; for example, continuity can be enforced at the DoFs located at crossing points of the domain decomposition.

Reviewer: ```suggestion
referred to as `primal`; for example, continuity can be enforced at the DoFs located at crossing points of the domain decomposition. Continuity of linear combinations of DoFs at the subdomain edge/faces, like averages, moments, and normal fluxes, is also possible, and may be needed depending on the choice of the underlying finite element space."
```

  header = ["Level", "DoFs", "Iterations", "Est. Condition Number"]
  print(f"\n{header[0]:<7} | {header[1]:<8} | {header[2]:<10} | {header[3]}")
  print("-" * 65)
  for row in results:
      print(f"{row[0]:<7} | {row[1]:<8} | {row[2]:<10} | {row[3]:.4f}")

Expected Outputs
----------------

When executed, the script outputs a table showing the scaling behavior of the
BDDC preconditioner. BDDC typically yields condition number bounds that grow
quasi-optimally as :math:`\mathcal{O}(1 + \log^2(H/h))`, where :math:`H` is
the subdomain size and :math:`h` is the mesh size. For cellwise subdomains
the ratio :math:`H/h` is equal to one.

Below is an example of what the performance output looks like when executed:

======== ======== ============ =====================
  Level    DoFs    Iterations  Est. Condition Number
======== ======== ============ =====================
0         441        20           5.6086
1         1681       21           5.7795
2         6561       21           5.8681
3         25921      21           5.9042
======== ======== ============ =====================

A runnable python version of this demo can be found :demo:`here

> <poisson_bddc.py>`.

Author: ```suggestion
<bddc_poisson.py>`.
```

      solver = LinearVariationalSolver(problem, solver_parameters=params)
      solver.solve()

      # Gather execution metrics
      iterations = solver.snes.getLinearSolveIterations()
      evals = solver.snes.ksp.computeEigenvalues().real
      kappa = max(evals, key=abs) / min(evals, key=abs)

      return V.dim(), iterations, kappa

Below we specify the BDDC solver parameters with cellwise subdomains,
note that this option is only available with a matrix-free operator. ::

  # Define the base direct solver configuration for the subproblems
  chol_params = {
      "pc_type": "cholesky",
      "pc_factor_mat_solver_type": "mumps",
  }

  # Solver parameters using firedrake.BDDCPC
  cellwise_bddc_params = {
      "mat_type": "matfree",
      "ksp_type": "cg",
      "ksp_norm_type": "natural",
      "ksp_converged_reason": None,
      "ksp_rtol": 1e-8,
      "ksp_atol": 0.0,

      # Enable tracking flags for condition numbers
      "ksp_view_eigenvalues": None,

      # Configure Firedrake's specialized BDDC Python Preconditioner
      "pc_type": "python",
      "pc_python_type": "firedrake.BDDCPC",
      "bddc_cellwise": True,

>       # Subproblem solvers
>       "bddc_pc_bddc_neumann": chol_params,
>       "bddc_pc_bddc_dirichlet": chol_params,
>       "bddc_pc_bddc_coarse": chol_params,

Reviewer: ```suggestion
      # Subproblem solvers
      "bddc_pc_bddc_neumann": chol_params,
      "bddc_pc_bddc_dirichlet": chol_params,
      "bddc_pc_bddc_coarse": chol_params,
      
      # Use only corner primal dofs
      "bddc_pc_bddc_use_edges" : False,
```

Reviewer: @pbrubeck You may want to rerun the example to collect the new iteration counts and condition numbers
